941 derive_param_wbc_abs

yangc71 opened the pull request:
Close #941 

Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://pharmaverse.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md`
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue so that it closes after successful merging. 
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

yangc71 commented:
Hi @millerg23 @bms63 , I created a first draft for the function, could you please take a look and let me know if the approach is ok? Also below are a few notes from myself regarding the current behavior of the function: 

- Currently, the function directly use `LBTESTCD` and `LBSTRESN`, and renames them to `PARAMCD` and `AVAL` in a temporary dataset within the function (so that I can leverage the `derive_derived_params()` function which only takes `PARAMCD` and `AVAL`). Will users ever derive the absolute lab parameters for WBC differentials based on `PARAMCD` and `AVAL` instead of the LB variables? If so, the function needs to be updated. 
- Do you think we actually need the `filter_diff` argument? I think we can just identify which records are the differentials based on `diff_code` supplied, but I'm not sure if there will be cases where users want to calculate absolute for only some of the records from `diff_code`. 
- I tried to keep the `by_vars` simple in my example, but I assume study data is more complicated (e.g. probably many lab tests are done in the same visit, and we need part of LBSPID to identify unique records/need to account for the RETEST ones etc.). I assume these will just be handled by the user in pre-processing steps and I do not need to incorporate them into this function? 

Thanks!

millerg23 commented:
> Hi @millerg23 @bms63 , I created a first draft for the function, could you please take a look and let me know if the approach is ok? Also below are a few notes from myself regarding the current behavior of the function:
> 
> * Currently, the function directly use `LBTESTCD` and `LBSTRESN`, and renames them to `PARAMCD` and `AVAL` in a temporary dataset within the function (so that I can leverage the `derive_derived_params()` function which only takes `PARAMCD` and `AVAL`). Will users ever derive the absolute lab parameters for WBC differentials based on `PARAMCD` and `AVAL` instead of the LB variables? If so, the function needs to be updated.
> * Do you think we actually need the `filter_diff` argument? I think we can just identify which records are the differentials based on `diff_code` supplied, but I'm not sure if there will be cases where users want to calculate absolute for only some of the records from `diff_code`.
> * I tried to keep the `by_vars` simple in my example, but I assume study data is more complicated (e.g. probably many lab tests are done in the same visit, and we need part of LBSPID to identify unique records/need to account for the RETEST ones etc.). I assume these will just be handled by the user in pre-processing steps and I do not need to incorporate them into this function?
> 
> Thanks!

Hi @yangc71,
I would expect function to use PARAMCD and AVAL, similar to the existing functions `derive_param_qtc`, `derive_param_map`, etc.
Regarding, `filter_diff`, I don't have a strong opinion on this, I can't think of a case to use it, so would be happy to remove it.
Regarding `by_vars`, I would agree, keep it fairly simple, its up to user to keep populate accurately based on their data. Ideally, we shouldn't be creating variables required for the BY group. I would expect users to create these themselves, via another function call.

millerg23 commented:
Hi @yangc71,
I had a quick look and have a couple of comments:
1) AVALU shouldn't be mentioned, it isn't in the AdAM IG, as the unit is part of PARAM. (Roche and I am sure other companies use it, but we are trying to follow CDISC as much as possible).
2) There is an {admiral} function that extracts the unit from PARAM, that you might need to use.
3) I would remove suffix of `SI` from `PARAMCD` and `(SI Result)` from `PARAM`, just to keep it simple. Other companies may not have these suffices.
4) I would cover dividing by zero, it shouldn't really happen in clean data, but output should be `NA` not `Inf`

yangc71 commented:
Hi @millerg23, thank you for your comments! I will update the function to be more general and consider users possibly using use PARAMCD and AVAL for calculation. 

Regarding your additional comments, I agree with 1 and 3 and will update the codes. For 2 and 4: 

- For 2, if I understand correctly I will need to add a `get_unit_expr` argument to the function (e.g. similar to the set up in `derive_param_bmi()`) that checks for the required unit of input parameters. But I wonder what unit should I set to? Are the units always going to be '10^9/L' for WBC and 'fraction of 1' for the differential? 
- For 4, can you please clarify where we are considering dividing by zero? I thought we would calculate the absolute value using the formula `Absolute Value = (WBC * Percentage Value) / 100` so if percent is 0 the result will just be 0. Please let me know if I misunderstood any part. 

Thanks!

yangc71 commented:
Hi @millerg23 @bms63 @konstantinakoukR, 
As discussed, I have made the following updates to the function and unit tests:
1. Removed reference to AVALU and SI unit.
2. Updated function to use PARAMCD and AVAL instead of LBTESTCD and LBSTRESN.
3. Removed `filter_diff`.
4. Added arguments to check units of input parameters. 
Please let me know if any further comments. Thanks!

millerg23 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,162 @@
+    filter(is.na(temp_flag))
+
+  # create new parameters.
+  temp_set_values_to <- quo_c(set_values_to, vars(temp_var = "new"))
```
do we need temporary variable `temp_var` created? Couldn't we use same filter on line 131 to filter on line 150? The code isn't wrong, just think we should avoid temporary variables if we can.

yangc71 replied:
No you are right, the `temp_var` was carried over with the previous code setup I had and I forgot to update this part. I have implemented your suggestion.

millerg23 commented on tests/testthat/test-derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,125 @@
+  "P01",    "WBC",     33,   "Leukocyte Count (10^9/L)",       "CYCLE 1 DAY 1",
+  "P01",    "WBC",     38,   "Leukocyte Count (10^9/L)",       "CYCLE 2 DAY 1",
+  "P01",    "LYMLE",   90,   "Lymphocytes (fraction of 1)", "CYCLE 1 DAY 1",
+  "P01",    "LYMLE",   70,   "Lymphocytes (fraction of 1)", "CYCLE 2 DAY 1"
```
AVAL for `"Lymphocytes (fraction of 1)"` should be 0.9, 0.7, etc.

yangc71 replied:
Updated.

bms63 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,162 @@
+#'
+#' @param dataset Input dataset
+#'
+#'   The variables specified by the `by_vars` argument, `PARAMCD`, and `AVAL`
```
Does anything happen if PARAMCD or AVAL are not present?  If so, should we state what happens?

yangc71 replied:
Yes `PARAMCD` and `AVAL` are checked in the `assert_data_frame()` call, and if those variables are not in the input dataset, an error e.g. "Required variable `PARAMCD` is missing" will be issued. I think it's sufficient to state the variables are required, because if they are not users will see the error.

bms63 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,162 @@
+#'   The result is used to check the units of the input parameters.
+#'
+#'   Permitted Values: a variable containing unit from the input dataset, or a function call
+#'   (for example, `get_unit_expr = extract_unit(PARAM)`).
```
Are the parentheses needed here?  
```suggestion
#'   , for example, `get_unit_expr = extract_unit(PARAM)`.
```

yangc71 replied:
Updated.

bms63 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,162 @@
+#'
+#' @param wbc_unit A string containing the required unit of the WBC parameter.
+#'
+#'   Default: `"10^9/L"`
```
I'm never sure when we should use quotes...if ever...Can you look over the Defaults of other program and see if they are present?

yangc71 replied:
I think we need quotes when the input value is a string, like this case here. If it's a variable name or expression then we don't need the quotes

bms63 replied:
thanks!

bms63 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,162 @@
+  assert_unit(dataset, wbc_code, required_unit = wbc_unit, get_unit_expr = !!get_unit_expr)
+  assert_unit(dataset, diff_code, required_unit = diff_unit, get_unit_expr = !!get_unit_expr)
+
+  # select observations and variables required for new observations.
```
Comments should start with Proper Case.  I think styler fixes things like this if you run it on the selection of code.

yangc71 replied:
Thanks, I capitalized the first word of the comments, or did you mean to capitalize each word?

bms63 replied:
Just the first word - I think it to sytler if fixes that kind of thing - but not sure

bms63 commented on tests/testthat/test-derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,125 @@
+test_that("Test adding absolute records for each by group", {
```
Do you mind writing out Test1:, Test2:, etc in your file.  The validation report appears slicker and more readable when we use this method.  Not widely adopted yet but I think our new adlb stuff should do this 
![image](https://user-images.githubusercontent.com/10111024/161298097-77cb97e8-9550-465d-ab92-d74709b564d1.png)
#714

yangc71 replied:
Updated.

yangc71 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,162 @@
+      parameters = c(wbc_code, diff_code),
+      by_vars = by_vars,
+      analysis_value = ((!!sym(paste0("AVAL.", wbc_code)) *
+                           !!sym(paste0("AVAL.", diff_code))) / 100),
```
I removed the `diff_unit` argument, and added a `diff_type` argument that takes in either `"fraction"` or `"percent"`. If `"fraction"` the function doesn't divide by 100, if "percent" it divides by 100. However, I kept the `wbc_unit` since it defaults to "10^9/L" which I think is the common unit? But if you think it's better to remove the unit checking I can do that too.

bms63 commented:
Hey Annie.  In the commit messages can you put #941 Thing done .  This way it links the change to the issue when we merge all this stuff in.

bms63 commented:
Are we just waiting on some feedback from Gordon?  Or is this ready to roll?  We should also ask Thomas or Stefan to take a look at it once we are happy with it.

yangc71 commented:
Thanks for the reminder @bms63! And I think we should wait for Gordon to take a look at the new changes. In the meantime, @thomas-neitmann @bundfussr would either of you be able to take a look to see if you have any comments? Thank you!

millerg23 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,172 @@
+#'
+#' @description
+#'
+#' Add a parameter by converting lab differentials from percentage to absolute values
```
Should be "from fraction or percentage to absolute values"

yangc71 replied:
Updated.

millerg23 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,172 @@
+#' @param diff_code white blood differential parameter
+#'
+#'   The observations where `PARAMCD` equals the specified value are considered
+#'   as the white blood differential lab results in percentage value to be converted
```
Should be "lab results in fraction or percentage value......"

yangc71 replied:
Updated.

millerg23 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,172 @@
+#' \deqn{\frac{White Blood Cell Count  * Percentage Value}{100}}
+#'
+#' If `diff_type` is `"fraction"`, the analysis value of the new parameter is derived as
+#' \deqn{White Blood Cell Count  * Percentage Value}
```
For `diff_type` is `"fraction"` should it be {White Blood Cell Count  * Fraction Value}

yangc71 replied:
Updated.

millerg23 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,172 @@
+      set_values_to = set_values_to
+    ) %>%
+    filter(PARAMCD == !!quo_get_expr(set_values_to$PARAMCD)) %>%
+    select(-starts_with("temp_"))
```
Could we drop "temp_flag" when dataset_temp created above ie. immediately after its used. It may be overkill but what if user has a TEMP_xxx variable on their input data set? Maybe I am over-thinking.....

millerg23 replied:
Can probably ignore what I said about "TEMP_xxx" variables, just saw this in https://pharmaverse.github.io/admiral/articles/programming_strategy.html
"If the input dataset includes variables starting with temp_, an error must be issued."

millerg23 reviewed (commented):
@yangc71 just some minor comments from me, looks really good. Unit tests look great.
@bms63 do we get a unit test score for each pull request? Related to another ticket where you are trying to get above 90% overall.

yangc71 commented on R/derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,172 @@
+#'
+#' @param diff_type A string specifying the type of differential
+#'
+#'   Permitted Values: `"percent"`, `"fraction"`
```
I agree and have updated the default to "fraction", this way users also have one less parameter to specify if they are using "fraction"...

yangc71 commented:
Thank you @millerg23! I have addressed the comments.

millerg23 reviewed (approved):
All good from my side, thanks Annie.

konstantinakoukR reviewed (approved):
Awesome function Annie! And great reviews provided. No further comments from me.

yangc71 commented:
Hi @bms63 @millerg23 @konstantinakoukR Thank you for your reviews! Just checking on how I should proceed with this PR? Are we still waiting on the check_template to be fixed or are there any other actions I need to take on my end?

bms63 commented on tests/testthat/test-derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,164 @@
+  )
+})
+
+test_that("TEST 3: Test when absolute record already present in source dataset 1", {
```
```suggestion
test_that("Test 3: Test when absolute record already present in source dataset 1", {
```

bms63 commented on tests/testthat/test-derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,164 @@
+})
+
+
+test_that("Test 3: Test when absolute record already present in source dataset 2", {
```
```suggestion
test_that("Test 4: Test when absolute record already present in source dataset 2", {
```

bms63 commented on tests/testthat/test-derive_param_wbc_abs.R:
```diff
@@ -0,0 +1,164 @@
+})
+
+
+test_that("Test 4: Test percent differential type", {
```
```suggestion
test_that("Test 5: Test percent differential type", {
```

bms63 reviewed (approved):
Hi Annie.  I made a couple of test suggestions.  Noticed the numbering got out of order.  I think we can get this into devel today.  Ping me when you make changes.

yangc71 commented:
Hi @bms63 , I updated the test number. However it seems the R-CMD-check for release is failing (it was working fine yesterday and the only thing I changed was the test number in this PR). It looks like for other PRs this is failing too? Could you please help take a look, thank you!

bms63 commented:
Hi Annie.  This is all good.  I brought in some of the changes for the checks.  Going to let them fire off and then i will merge into devel by EOD.  Thanks!!

bms63 commented:
Everything seems to check out!  Going to merge in!  Great job again Annie!!  This is a really nice addition to our suite of lab functions.
